### PR TITLE
Create 'immutable artefacts' pattern

### PR DESCRIPTION
**Pull request by edhamiltonHO**
# Content change 

I can confirm:
- [x] Content does not include any code or configuration changes (excluding frontmatter information)
- [x] Content meets the content standards
e.g. [Writing a principle](https://ho-cto.github.io/engineering-guidance-and-standards/docs/standards/writing-a-principle/) and [Writing a standard](https://ho-cto.github.io/engineering-guidance-and-standards/docs/standards/writing-a-standard/)
- [x] Content is suitable to open source, i.e.:
    - Content does not relate to unreleased gov policy
    - Content does not refer to anti-fraud mechanisms
    - Content does not include sensitive business logic

### CODE REVIEW

File: docs/patterns/immutable-artefacts.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+- disaster recovery (consider rebuilding an environment from scratch)
+
+You should:
+
+- Produce an immutable artefact early in your build pipeline and store it with a unique version in a repository. This should be automated where possible. For example:
+```
+${semver}-${build-number}-${commit-hash-fragment}
+```
+- Promote this artefact through each assurance step and link the results to the artefact to provide an audit record.
+- Keep configuration out of the artefact. This should be provided using environment variables or similar external methods. See also Store Infrastructure and Configuration as Code.
```
Comment (aaronrussellHO) [inline docs/patterns/immutable-artefacts.md:39]: Can this be a link to the standard for this?

File: docs/patterns/immutable-artefacts.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+- Deployment
+- Artefacts
+---
+
+{% import "macros/tags.njk" as pageTag %}
+
+Last updated: {{ page.date | postDate }}
+{{ pageTag.tags(tags)  }}
+
+An artefact is the deployable unit of code, for instance a Docker image, serverless package, or virtual machine image. You should build artefacts once, that you never modify, and use them across all environments. In other words, if you need to make changes, a new artefact should be created. This helps with assurance and debugging incidents.
```
Comment (aaronrussellHO) [inline docs/patterns/immutable-artefacts.md:17]:
```suggestion
An artefact is the deployable unit of code, for instance a Docker image, serverless package, or virtual machine image. You should build artefacts once, afterwards and you should never modify them. If you need to make changes, a new artefact should be created. This helps with assurance and debugging incidents. 

An artefact should then be promoted between environments, this enables you to know what is running as the same arefact and code is used across different environments.
```
Think this reads a little clunky, how's this?

File: docs/patterns/immutable-artefacts.md [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+With proper versioning of your artefacts, you will be able to tie your environments back to a specific commit in your code-base, which is useful for investigation of incidents. Without immutable artefacts you may have hidden state and/or subtle differences between environments, that make it very hard to understand the behaviour your are observing and undermine your confidence that something will continue to work when moved to another environment. You will also find it hard to recover from a future disaster, as you will need to rebuild your artefact first, and your build pipeline may be broken.
```
Comment (aaronrussellHO) [inline docs/patterns/immutable-artefacts.md:19]:
```suggestion
With proper versioning of your artefacts, you will be able to tie your environments back to a specific commit in your code-base, which is useful for investigation of incidents. Without immutable artefacts you may have hidden state and/or subtle differences between environments, that make it very hard to understand the behaviour you are observing and undermine your confidence that something will continue to work when moved to another environment. You will also find it hard to recover from a future disaster, as you will need to rebuild your artefact first, and your build pipeline may be broken.
```

File: docs/patterns/immutable-artefacts.md [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+---
+
+## Solution
+
+Immutable artefacts help us to know what we are running. They aid us in accurately reproducing environments and so help with:
```
Comment (aaronrussellHO) [inline docs/patterns/immutable-artefacts.md:25]:
```suggestion
Immutable artefacts help us to know what we are running. They aid us in accurately reproducing environments, this helps with:
```

File: docs/patterns/immutable-artefacts.md [hunk 5]
```diff
@@ -0,0 +1,41 @@
+---
+layout: sub-navigation
+order: 1
+title: Immutable artefacts
+date: git Last Modified
+tags:
+- Build, release and deploy
+- Deployment
```
Comment (aaronrussellHO) [inline docs/patterns/immutable-artefacts.md:8]: Does it need a patterns tag?

Comment (edhamiltonHO) [reply docs/patterns/immutable-artefacts.md:8]: Patterns tag is automatically applied during 11ty process, by virtue of being in the patterns subdirectory 🙂

File: docs/patterns/immutable-artefacts.md (hunk 1 above)
Comment (edhamiltonHO) [reply docs/patterns/immutable-artefacts.md:39]: Yes, can update this as soon as the infrastructure as code standard is merged

File: docs/patterns/immutable-artefacts.md (hunk 2 above)
Comment (edhamiltonHO) [reply docs/patterns/immutable-artefacts.md]: Have rewritten this entirely to be clearer on what an immutable artefact is

**Review by aaronrussellHO — APPROVED** [review]

File: docs/patterns/immutable-artefacts.md [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+## Solution
+
+You should follow a '[build once, deploy many](https://www.mikemcgarr.com/blog/build-once-deploy-many.html)' approach. Specifically, you should:
+
+- When you create a new version of your application or service (typically, in response to merging some new code into your main branch), build your application once and package it into an artefact along with its dependencies
+- Store artefacts in a suitable artefact repository with a unique version identifier. Versioning should follow a common pattern, and be automated where possible. For example:
+```
+${semver}-${build-number}-${commit-hash-fragment}
+```
```
Comment (ezramorris) [inline docs/patterns/immutable-artefacts.md:38]:
```suggestion
- Store artefacts in a suitable artefact repository with a unique version identifier. Versioning should follow a common pattern, and be automated where possible, for example using [Semantic Versioning](https://semver.org/) or simply the build number.
```

I think the example is a bit confusing. Suggest referring to SemVer; it covers things like pre-release builds.

**Review by ezramorris — CHANGES_REQUESTED** [review]

**Review by ezramorris — APPROVED** [review]

File: docs/patterns/immutable-artefacts.md (hunk 1 above)
Comment (edhamiltonHO) [reply docs/patterns/immutable-artefacts.md]: Now that the infrastructure and configuration as code standard is focussed exclusively on infrastructure as code this link is not relevant any more

**Pull request merged**